Add dnssd support to cpu

ericvh opened the pull request:
First draft of mdns support to cpu and cpud

This is still a WIP, but is functional - it just selects the first mdns response which I want to expand to allow for selection criteria and prioritization criteria to allow the user to say they want to CPU into the lowest-load arm based cpu-server.

I probably need some help figuring out the right incantations to incorporate dependencies into the .circleci build - my build is based on my Dockerfile and .gitlab-ci.yml

Signed-off-by: Eric Van Hensbergen <eric.vanhensbergen@arm.com>

ericvh committed:
Add optional mdns support to cpu/cpud

Parameters to cpud enable mdns and set optional TXT key=value fields.
Parameters to cpu use mdns to select host and use TXT key=value match to select and prioritize.

Signed-off-by: Eric Van Hensbergen <eric.vanhensbergen@arm.com>

rminnich commented on cmds/cpud/serve.go:
```diff
@@ -91,14 +95,20 @@ func listen(network, port string) (net.Listener, error) {
 		}
 		ln, err = vsock.ListenContextID(any, uint32(p), nil)
-
+		if err != nil {
```
This return is not needed. 

There seem to be several different things going on in this PR, can you split out into multiple PRs. It makes bisect way easier.

rminnich commented on cmds/cpud/serve.go:
```diff
@@ -115,6 +125,31 @@ func serve() error {
 	if err != nil {
 		return err
 	}
+	// TODO: This is overly restrictive
```
This should be done in the server package.

rminnich replied:
Could we indicate mdns in the network address somehow and let client code figure it out? 
What if as part of hostname parsing we could just magically now on the client side what to do? What if on server side we could, similarly, know that cpud needs to be an mdns server? There seem to be some kind of standards out there for it. I really want to get this in, it would be super handy, just trying to work out how. 

Can we just mount an mdns server on /net/cs -- oh wait ...

ericvh replied:
Yeah, I did think that we could do magic hostname specification -- of course that always comes with its own caveats.  Server side autodetecting if it has to act as server seems a bit more dodgey - in general things seem to work on metal, but docker won't let multiple people control the port.  There seem to be multiple things out there to make docker play nice with mdns, but haven't had time to really try any of them in anger.  In any case, probably will need to make that different variations of options (host your own server, play nice with dbus, use our own mdns port, do something else)

Sorry - was a busy week this last week, so I didn't get around to doing much here - was planning on spending some time this week implementing the code reorg into the modules like you suggested and fleshing out the meta-data based masking and prioritization.

rminnich replied:
very cool, I was surprised to see this closed but if you're going to keep at it, we REALLY need it. It's going to make life a lot better for our use cases.

ericvh replied:
server portion migrated to server module, but it wasn't as clean as I would like.  Suggestions welcome.

rminnich commented on cmds/cpu/cpu.go:
```diff
@@ -170,12 +175,32 @@ func newCPU(host string, args ...string) error {
-		usage()
+	a := ""
+
+	if *mdnsEnabled {
```
This probably needs to be in the client package.

ericvh replied:
@rminnich - pushed movement of mdns into client module -- this relies on you being able to set network to mdns and using hostname to specify selection/prioritization criteria. This pass muster or do you want me to try and parse URI to specify mdns -- and then what kind of syntax:
mdns://arch=arm64,mem>16G:cpucount>16
or
mdns://_ncpu._tcp(arch=arm64,mem>16G:cpucount>16)
or
mdns!_ncpu._tcp!_arch=arm64,mem>16G:cpucount>16
(or something else)

Of course, this may be the next level of abstraction - maybe just mdns to start with as hostname?

rminnich reviewed (commented):
There's def one thing you don't need here. Let's start with that.

The way this is done will make it hard to write tests, so let's figure that out too. We don't have enough coverage as it is, so we don't want to make  it worse.

rminnich commented on cmds/cpud/serve.go:
```diff
@@ -91,14 +95,20 @@ func listen(network, port string) (net.Listener, error) {
 	default:
 		ln, err = net.Listen(network, net.JoinHostPort("", port))
+		if err != nil {
+			return nil, err
```
This test is not needed.

ericvh replied:
removed

rminnich commented on cmds/cpud/serve.go:
```diff
@@ -91,14 +95,20 @@ func listen(network, port string) (net.Listener, error) {
+		}
+
 	}
 	return ln, err
```
Evreryone returns here.

rminnich commented on cmds/cpud/serve.go:
```diff
@@ -115,6 +118,31 @@ func serve() error {
+		// Setup our service export
+		host, _ := os.Hostname()
+		var p uint64
+		p, err = strconv.ParseUint(*port, 10, 32)
```
Are ports 32 bits? Do we need to require base 10? People might want to say port 0xcafe maybe?

ericvh replied:
Agreed.  will stop specifying the port conversion.

abrender commented on server/server.go:
```diff
@@ -105,6 +108,28 @@ func handler(s ssh.Session) {
+func Advertise(hostname, port, txt string) (*mdns.Server, error) {
+	v("starting mdns server")
+	// Setup our service export
+	host, _ := os.Hostname()
```
Is it safe to ignore the error here? I think you might want to handle it?

abrender replied:
(Specifically referring to the error returned by `os.Hostname()`)

ericvh replied:
It's a fair comment.  In some ways I was trying to follow lighter weight error handling focus elsewhere in the code -- but this sort of error likely to cause issues so maybe better explicitly report and then bail out.

abrender commented on server/server.go:
```diff
@@ -105,6 +108,28 @@ func handler(s ssh.Session) {
+		server, err := mdns.NewServer(&mdns.Config{Zone: service})
+		return server, err
+	} else {
+		log.Printf(`mdns new service failure: %v`, err)
```
Do you want to return an error here instead of logging and returning err = nil?

I'm trying to understand how the code is structured - is the goal here to still return a nil error if mdns.NewMDNSService() fails?    If not then maybe you can simplify and restructure the code as:
```
service, err := ....
if err != nil {
  // Handle the error
}
return mdns.NewServer(&mdns.Config{Zone: service})
```
?

ericvh replied:
the original goal was to try and keep mdns as optional as possible so if mdns resolution fails, just fall through and hope the user specified the hostname.  I figured the end result would be a host not found if otherwise -- but I'm happy to take action to make this more explicit or specific (no available resource found versus host not found)

abrender replied:
Thanks for letting me know the goal, that is helpful.   In this case I would think that it's still better to explicitly return the error here from `Advertise` in server.go (because we are failing to advertise...), but ignore the error in the *caller* of the function, which appears to be cpud/serve.go in the `if *mdnsEnabled{` section of code - and it appears you're already doing that.

So properly returning the errors here won't change the existing optionality that is already accomplished by the caller in cpud/serve.go ignore the returned error.

Please let me know if I've misunderstood? Thanks!

ericvh replied:
I've added back in error handling and made failure less silent if you selected the mdns advertisement option.

abrender commented on cmds/cpud/serve.go:
```diff
@@ -115,6 +114,12 @@ func serve() error {
+
+	if *mdnsEnabled {
+		host, _ := os.Hostname()
+		mdns_server, _ := server.Advertise(host, *port, *mdnsTxt)
```
I understand that the intent here is to make this optional so the error is ignored. But future readers of the code might not understand this, so can you please add a comment about why the error is ignored? Thanks!

abrender commented on cmds/cpud/serve.go:
```diff
@@ -115,6 +114,12 @@ func serve() error {
+	if *mdnsEnabled {
+		host, _ := os.Hostname()
+		mdns_server, _ := server.Advertise(host, *port, *mdnsTxt)
+		defer mdns_server.Shutdown()
```
calling `mdns_server.Shutdown()` probably won't work if `server.Advertise` returned an error, right? So maybe you can structure the code as

```
// Comment here explaining why we will ignore the `err != nil` case.
if mdns_server, err := server.Advertise(...); err == nil {
  defer mdns_server.Shutdown()
}
```

ericvh replied:
actually, reflecting on this a bit more, if the user specifies mdns advertisment on the command line and it goes sideways, I think we just treat it as failure.  As long as mdns advertisement isn't default, this should be fine.  I'll restructure code appropriately.

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+		txt[s[0]] = s[1]
+	}
+	for k, _ := range req {
+		if !contains(req[k], txt[k]) {
```
I think you can use https://pkg.go.dev/golang.org/x/exp/slices#Contains here.  It is my understanding that this is quite possibly going to be merged into the standard library in the future.

Note that both the above package and the current implementation are O(N) which is expensive... if you use a map then it's O(1).  I don't know the size of these slices to know if that's important enough though... so I think the performance implications of the slower implementation should be ignored in favor of the simplicity of using https://pkg.go.dev/golang.org/x/exp/slices#Contains

ericvh replied:
Thanks for the pointer.  This shouldn't be in a performance critical piece of code, so it shouldn't be a huge issue.

ericvh replied:
Had to update the underlying image for CI, but was able to follow your suggestions.

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+
+	// determine if we want to do mdns
+	u, err := url.Parse(uri)
+	if (err == nil) && (u.Scheme == "dnssd") {
```
you can save an entire layer of indent for the following ~30 lines of code if you structure the code as:
```
if err != nil {
  // return code
}
if u.Scheme != "dnssd") {
  return "", "", "", nil // Nothing to do here...
}
// rest of code.
```

ericvh replied:
yeah, I waffled between the two, happy to side with not having the indent.

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+
+		// Make a channel for results and start listening
+		entriesCh := make(chan *mdns.ServiceEntry, 64)
+		// Setup sensible defaults"
```
Is the quotation mark at the end of this comment a typo?

ericvh replied:
well spotted

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+		mdns_txt := u.Query()
+
+		// Make a channel for results and start listening
+		entriesCh := make(chan *mdns.ServiceEntry, 64)
```
Where does the number 64 come from?

ericvh replied:
its arbitrary, nearly as arbitrary as the 4 that was there before -- felt like a deeper channel would be less likely to miss options, happy to have suggestions on best practices here.

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+		// Setup sensible defaults"
+		// Start the lookup
+		err = mdns.Lookup(mdns_type+"."+mdns_domain, entriesCh)
+		if err != nil {
```
I'm confused by this, should it be `if err == nil {` ? The following code ranges through the entriesCh channel, but if `err != nil` then I'd not expect this to work?

ericvh replied:
well spotted.  I committed and pushed at the end of the day once things compiled so I could test at home (to be fair, the Pull request is still marked as WIP) -- thanks for early debugging this ;)

abrender commented on client/client.go:
```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
+				if mdns_required(entry.InfoFields, mdns_txt) {
+					hostname = entry.Addr.String()
+					port = strconv.Itoa(entry.Port)
+					network = "tcp"
```
is `network` the right name for this variable? Should it be `protocol`?

ericvh replied:
I was following the paradigm of the client options which callas it network -- I do realize that its unlikely we'd actually get a unix or a vsock response back from mdns...but those are more network than protocol.  It starts to get fuzzy.

rminnich replied:
it's a go thing ... they use net and network when they kind of mean protocol. This is due to Plan 9 root IIUC.

abrender commented on cmds/cpud/serve.go:
```diff
@@ -115,6 +116,15 @@ func serve() error {
+		mdns_server, err := server.Advertise(*port, *mdnsTxt)
+		if err != nil {
+			return fmt.Errorf("Could not advertise with mdns: %w", err)
+		} else {
```
I don't think this `else` is needed because the previous `if` statement `return`s
```
if err != nil {
  return fmt.Errorf(....)
}
defer mdns_server.Shutdown()
```

ericvh replied:
true, will simplify.

abrender commented on server/server.go:
```diff
@@ -105,6 +108,27 @@ func handler(s ssh.Session) {
+		return nil, fmt.Errorf("Could not instantiate mDNS server on port %d: %w", p, err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: service})
```
You can do `return mdns.NewServer(...)` here instead of `server, err := ...; return server, err`

ericvh replied:
will simplify.

ericvh commented:
There are some unfortunate problems due to the lack of functionality (one might even say bugs) in the hashicorp/mdns package. I'm currently investigating swapping it out and will also look at incorporating the experimental slice compare library.

ericvh commented:
Okay, I'm fairly happy with this as MVP -- let me know if you want me to do a rebase/squash to make it easier to review.

ericvh committed:
cleanup transitivty of verbose

Added some functions to allow CLI to set verbose flags in module.
Cleaned these up to look the same in client and server and generalized
them to let calling module set the logging function.

Signed-off-by: Eric Van Hensbergen <eric.vanhensbergen@arm.com>
